runtime/stdlib/account.go

```diff
@@ -670,21 +678,117 @@ func newAccountKeysGetFunction(
 					inter,
 					locationRange,
 					accountKey,
-					// public keys are assumed to be already validated.
-					func(
-						_ *interpreter.Interpreter,
-						_ interpreter.LocationRange,
-						_ *interpreter.CompositeValue,
-					) error {
-						return nil
-					},
+					assumePublicKeyIsValid,
 				),
 			)
 		},
 		sema.AccountKeysTypeGetFunctionType,
 	)
 }
 
+func newAccountKeysForEachFunction(
+	gauge common.MemoryGauge,
+	provider AccountKeyProvider,
+	addressValue interpreter.AddressValue,
+) *interpreter.HostFunctionValue {
+	address := addressValue.ToAddress()
+	// the AccountKey in `forEachKey(_ f: ((AccountKey): Bool)): Void`
+	callbackTypeParams := []sema.Type{sema.AccountKeyType}
+
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			fnValue, ok := invocation.Arguments[0].(interpreter.FunctionValue)
+
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			inter := invocation.Interpreter
+			locationRange := invocation.LocationRange
+
+			newSubInvocation := func(key interpreter.Value) interpreter.Invocation {
+				return interpreter.NewInvocation(
+					inter,
+					nil,
+					[]interpreter.Value{key},
+					callbackTypeParams,
+					nil,
+					locationRange,
+				)
+			}
+
+			liftKeyToValue := func(key *AccountKey) interpreter.Value {
+				return NewAccountKeyValue(
+					inter,
+					locationRange,
+					key,
+					assumePublicKeyIsValid,
+				)
+			}
+
+			count := provider.AccountKeysCount(address)
+
+			var err error
+			var accountKey *AccountKey
+
+			var index uint64
+			for index = 0; index < count; index++ {
+				wrapPanic(func() {
+					accountKey, err = provider.GetAccountKey(address, int(index))
+				})
+				if err != nil {
+					// Here it is expected the host function to return a nil key, if a key is not found at the given index.
+					// This is done because, if the host function returns an error when a key is not found, then
+					// currently there's no way to distinguish between a 'key not found error' vs other internal errors.
```

Comment:
This should go above the if-statement below, like in `newAccountKeysGetFunction`

Reply:
ah yeah, been trying to figure out the failing test to no avail. meeting with @dsainati1 tomorrow to discuss it.

the failure stems from `exportValueWithInterpreter`, where the runtime is attempting to resolve a `count` field on the parent AuthAccount's `inbox` type. I think scoping and/or aliasing is causing the inbox type's members to be mutated during repeated nested calls to `exportCompositeType`, but idk where exactly.

Reply:
Huh? Very odd the inbox changes related to this. I'll have a look as well

Reply:
The Account Inbox PR #1983 had a couple mistakes, I'll open a PR to fix them and add tests to prevent future similar ones

Reply:
Opened #2053 which fixes the mistakes and unblocks this PR. Let's get that one in first, and then once it is in, we can merge master into this and CI should pass (it does locally). There's nothing left to do for this PR, it should resolve itself. Please have a look at the bugfix PR though.